client/state/data-layer/http-data.ts

```diff
@@ -0,0 +1,345 @@
+	const lastUpdated: TimestampMS = Date.now();
+	const item = httpData.get( id ) || empty;
+
+	// We could have left out the keys for
+	// the previous properties if they didn't
+	// exist but I wanted to make sure we can
+	// get our hidden classes to optimize here.
+	switch ( state ) {
+		case DataState.Failure:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: data,
+				lastUpdated: item.lastUpdated,
+				pendingSince: undefined,
+			} );
+
+		case DataState.Pending:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: undefined,
+				lastUpdated: item.lastUpdated,
+				pendingSince: lastUpdated,
+			} );
+
+		case DataState.Success:
+			return httpData.set( id, {
+				state,
+				data,
+				error: undefined,
+				lastUpdated,
+				pendingSince: undefined,
+			} );
+	}
+};
+
+export const update = ( id: DataId, state: DataState, data?: unknown ) => {
+	const updated = updateData( id, state, data );
+
+	listeners.forEach( f => f() );
+
+	return updated;
+};
+
+const fetch = ( action: HttpDataAction ) => {
+	update( action.id, DataState.Pending );
+
+	return [
+		{
+			...action.fetch,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
+		},
+		{ type: HTTP_DATA_TICK },
+	];
+};
+
+const onError = ( action: HttpDataAction, error: any ) => {
+	update( action.id, DataState.Failure, error );
+
+	return { type: HTTP_DATA_TICK };
+};
+
+type SuccessfulParse = [undefined, ReturnType< ResponseParser >];
+type FailedParse = [Error, undefined];
+type ParseResult = SuccessfulParse | FailedParse;
+
+/**
+ * Transforms API response data into storable data
+ * Returns pairs of data ids and data plus an error indicator
+ *
+ * [ error?, [ [ id, data ], [ id, data ], … ] ]
+ *
+ * @example:
+ *   --input--
+ *   { data: { sites: {
+ *     14: { is_active: true, name: 'foo' },
+ *     19: { is_active: false, name: 'bar' }
+ *   } } }
+ *
+ *   --output--
+ *   [ [ 'site-names-14', 'foo' ] ]
+ *
+ * @param data - input data from API response
+ * @param fromApi - transforms API response data
+ * @return output data to store
+ */
+const parseResponse = ( data: any, fromApi: ResponseParser ): ParseResult => {
+	try {
+		return [ undefined, fromApi( data ) ];
+	} catch ( error ) {
+		return [ error, undefined ];
+	}
+};
+
+const onSuccess = ( action: HttpDataAction, apiData: unknown ) => {
+	const [ error, data ] = parseResponse( apiData, action.fromApi() );
+
+	if ( undefined !== error ) {
+		return onError( action, error );
+	}
+
+	update( action.id, DataState.Success, apiData );
+	( data as ResourcePair[] ).forEach( ( [ id, resource ] ) =>
+		update( id, DataState.Success, resource )
+	);
+
+	return { type: HTTP_DATA_TICK };
+};
+
+export default {
+	[ HTTP_DATA_REQUEST ]: [
+		dispatchRequest( {
+			fetch,
+			onSuccess,
+			onError,
+		} ),
+	],
+};
+
+export const reducer: Reducer< number > = ( state = 0, { type } ) =>
```

Comment:
Can we gain anything from typing the second arg as `Action< typeof HTTP_DATA_TICK >`? (Am I doing this right? :sweat_smile: )

Reply:
Oh you're doing it right, and we could. The Reducer type accepts an action type:

```suggestion
export const reducer: Reducer< number, Action< typeof HTTP_DATA_TICK > > = ( state = 0, { type } ) =>
```